pkg/grpc/api.go

```diff
@@ -407,13 +408,20 @@ type invokeServiceResp struct {
 	trailers metadata.MD
 }
 
+var (
+	invokeServiceDeprecationNoticeShown     = atomic.Bool{}
+	invokeServiceHTTPDeprecationNoticeShown = atomic.Bool{}
+)
+
 // Deprecated: Use proxy mode service invocation instead.
 func (a *api) InvokeService(ctx context.Context, in *runtimev1pb.InvokeServiceRequest) (*commonv1pb.InvokeResponse, error) {
 	if a.directMessaging == nil {
 		return nil, status.Errorf(codes.Internal, messages.ErrDirectInvokeNotReady)
 	}
 
-	apiServerLogger.Warn("[DEPRECATION NOTICE] InvokeService is deprecated and will be removed in the future, please use proxy mode instead.")
+	if invokeServiceDeprecationNoticeShown.CompareAndSwap(false, true) {
```

Comment:
If you want.

The two are similar but sync.Do is a bit slower and the reason why is explained in the source code of sync.Do itself:

https://github.com/golang/go/blob/245e95dfabd77f337373bf2d6bb47cd353ad8d74/src/sync/once.go#L49-L61

The issue explained in the comments shouldn’t apply to our case due to the fact that we’re just printing a log

Reply:
I'd rather for readability, even if this entire code path goes goes away in 1.12.

Reply:
Changed. Will trigger new perf tests as this will cause all parallel requests to be blocked until the first one has written the log.

Reply:
Looks like the perf tests for service_invocation_grpc are once again failing after making this change. I think this is plausible since with this change, requests that are coming in at the same time as the first one are blocked.

Ok to revert?

Reply:
If it's just for readability how about a combination of performance and readability: an utility `once` function that returns a function with a `no-wait` behavior?

```golang
var logDeprecationWarnOnce = utils.OnceNoWait(logDeprecationWarn)

func...

     logDeprecationWarnOnce()
```

Reply:
I reverted the change and added a lengthy explanation in the comments.